xarray/core/dataset.py

```diff
@@ -3043,20 +3048,45 @@ def _validate_interp_indexer(x, new_x):
             else:
                 use_indexers = validated_indexers
 
-            if var.dtype.kind in "uifc":
+            dtype_kind = var.dtype.kind
+            if dtype_kind in "uifc":
+                # For normal number types do the interpolation:
                 var_indexers = {k: v for k, v in use_indexers.items() if k in var.dims}
                 variables[name] = missing.interp(var, var_indexers, method, **kwargs)
+            elif dtype_kind in "ObU" and (use_indexers.keys() & var.dims):
+                # For types that we do not understand do stepwise
+                # interpolation to avoid modifying the elements.
+                # Use reindex_variables instead because it supports
+                # booleans and objects and retains the dtype but inside
+                # this loop there might be some duplicate code that slows it
+                # down, therefore collect these signals and run it later:
+                to_reindex[name] = var
             elif all(d not in indexers for d in var.dims):
-                # keep unrelated object array
+                # For anything else we can only keep variables if they
+                # are not dependent on any coords that are being
+                # interpolated along:
                 variables[name] = var
 
+        if to_reindex:
+            # Reindex variables:
+            variables_reindex = alignment.reindex_variables(
+                variables=to_reindex,
+                sizes=obj.sizes,
+                indexes=obj.xindexes,
```

Comment:
@benbovy maybe you've noticed this too but `indexes=obj.xindexes` isn't that intuitive and doing `indexes=obj.indexes` doesn't work anymore (see the xindexes commits). Not sure if `alignment.reindex_variables` is public though.

Reply:
@Illviljan sorry for the long wait. Yes `xindexes` is not really documented yet (there's only some docstrings for now) and `xindexes` vs `indexes` is confusing indeed (this will be fixed soon). This should only affect Xarray internals, though.